tck/src/main/java/org/eclipse/microprofile/lra/tck/TckTestBase.java

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    @Before
+    public void before() {
+        LOGGER.info("Running test: " + testName.getMethodName());
+        setUpTestCase();
+
+        try {
+            tckSuiteTarget = tckSuiteClient.target(URI.create(new URL(config.tckSuiteBaseUrl()).toExternalForm()));
+            lraClient = new LRAClientOps(tckSuiteTarget);
+        } catch (MalformedURLException mfe) {
+            throw new IllegalStateException("Cannot create URL for the LRA TCK suite base url " + config.tckSuiteBaseUrl(), mfe);
+        }
+        recoveryTarget = recoveryCoordinatorClient.target(URI.create(recoveryCoordinatorBaseUrl.toExternalForm()));
+    }
+
+    /**
+     * Checking if coordinator is running, set ups the client to contact the recovery manager and the TCK suite itself.
+     */
+    private void setUpTestCase() {
+        if(recoveryCoordinatorBaseUrl != null) {
+            // we've already set up the recovery urls and REST clients for the tests
+            return;
+        }
+
+        try {
+            // see https://github.com/eclipse/microprofile-lra/issues/116
+            recoveryCoordinatorBaseUrl = new URL(String.format("http://%s:%d/%s",
+                    config.recoveryHostName(), config.recoveryPort(), config.recoveryPath()));
+
+            tckSuiteClient = ClientBuilder.newClient();
+            recoveryCoordinatorClient = ClientBuilder.newClient();
+        } catch (MalformedURLException e) {
+            throw new RuntimeException("Cannot properly setup the TCK tests (coordinator endpoint, testsuite endpoints...)", e);
+        }
+    }
+
+    private void checkStatusAndCloseResponse(Response.Status expectedStatus, Response response, WebTarget resourcePath) {
+        try {
+            assertEquals("Not expected status at call '" + resourcePath.getUri() + "'",
+                    expectedStatus.getStatusCode(), response.getStatus());
+        } finally {
+            response.close();
+        }
+    }
+
+    String checkStatusReadAndCloseResponse(Response.Status expectedStatus, Response response, WebTarget resourcePath) {
+        try {
+            assertEquals("Response status on call to '" + resourcePath.getUri() + "' failed to match.",
+                    expectedStatus.getStatusCode(), response.getStatus());
+            return response.readEntity(String.class);
+        } finally {
+            response.close();
+        }
+    }
+
+    /**
+     * The started LRA will be named based on the class name and the running test name.
+     */
+    protected String lraClientId() {
+        return this.getClass().getSimpleName() + "#" + testName.getMethodName();
+    }
+
+    /**
+     * Adjusting the default timeout by the specified timeout factor
+     * which can be defined by user.
+     */
+    protected long lraTimeout() {
+        return Util.adjust(LraTckConfigBean.LRA_TIMEOUT_MILLIS, config.timeoutFactor());
+    }
+
+
+    void replayEndPhase(String resource) {
```

Comment:
Exactly, there should be an option to recover for a singe resource - that's why I added the parameter. But we don't need a schema, the resource to recover can be inferred from the location of the RECOVER annotation if we can get some agreement on that.

Reply:
This feature is still up for discussion. With my proposal on issue #116 it is per resource (but could be for all resources if required). BTW you said that you do not need the recovery so that issue will probably be moved out to the interoperability section of  milestone 1.x. When/if that decision is made I will rewrite the replayEndPhase functionality to use a blocking wait for the implementation to get around to retrying the end phase.

The TCK already uses a proprietary technique and my plan is to leave this functionality as is until we decide what to do with #116 and like I said that will probably end up being an indefinite blocking call - I don't want to implement something in this PR and then find that I have rewrite it again depending on the outcome of the discussion.

Reply:
@rdebusscher I know we discussed in the hangout that I would propose a PR for the recovery semantics and we would use that as a basis for discussion. But since I am the only person that thinks the RECOVERY annotation is useful I doubt it will make it into the spec (and we will need to fall back to something like "... after a crash the implementation must trigger recover at its earliest convenience ..."